File: src/main/scala/com/kotori316/fluidtank/recipes/FluidTankConditions.java
package com.kotori316.fluidtank.recipes;

import java.util.function.Predicate;

import com.google.gson.JsonObject;
import net.fabricmc.fabric.api.resource.conditions.v1.ConditionJsonProvider;
import net.minecraft.resources.ResourceLocation;

import com.kotori316.fluidtank.FluidTank;

public abstract class FluidTankConditions<T extends FluidTankConditions<T>> implements Predicate<JsonObject> {
    public FluidTankConditions(ResourceLocation location) {
        this.location = location;
    }

    private final ResourceLocation location;
    public final ConditionJsonProvider serializer = new Serializer();

    public ResourceLocation getID() {
        return this.location;
    }

    @Override
    public abstract boolean test(JsonObject context);

    private class Serializer implements ConditionJsonProvider {
        @Override
        public void writeParameters(JsonObject object) {
        }

        @Override
        public ResourceLocation getConditionId() {
            return FluidTankConditions.this.getID();
        }

    }

    public static final class ConfigCondition extends FluidTankConditions<ConfigCondition> {

        public static final ResourceLocation LOCATION = new ResourceLocation(FluidTank.modID, "config");

        public ConfigCondition() {
            super(LOCATION);
        }

        @Override
        public boolean test(JsonObject context) {
            return FluidTank.config.enableUpdateRecipe;
        }
    }

    public static final class EasyCondition extends FluidTankConditions<EasyCondition> {

        public static final ResourceLocation LOCATION = new ResourceLocation(FluidTank.modID, "easy");

        public EasyCondition() {
            super(LOCATION);
        }

        @Override
        public boolean test(JsonObject context) {
            return FluidTank.config.easyRecipe;
        }
    }

}

File: src/main/scala/com/kotori316/fluidtank/recipes/FluidTankConditions.scala
package com.kotori316.fluidtank.recipes

import com.google.gson.JsonObject
import com.kotori316.fluidtank.{Config, FluidTank}
import net.minecraft.resources.ResourceLocation
import net.minecraftforge.common.crafting.conditions.{ICondition, IConditionSerializer}

object FluidTankConditions {
  final class ConfigCondition extends FluidTankConditions[FluidTankConditions.ConfigCondition](new ResourceLocation(FluidTank.modID, "config")) {
    override def test: Boolean = !Config.content.removeRecipe.get()
  }

  final class EasyCondition extends FluidTankConditions[FluidTankConditions.EasyCondition](new ResourceLocation(FluidTank.modID, "easy")) {
    override def test: Boolean = Config.content.easyRecipe.get()
  }

  final class Serializer[T <: FluidTankConditions[T]](recipe: FluidTankConditions[T]) extends IConditionSerializer[FluidTankConditions[T]] {
    override def write(json: JsonObject, value: FluidTankConditions[T]): Unit = {
    }

    override def read(json: JsonObject): FluidTankConditions[T] = recipe

    override def getID: ResourceLocation = recipe.getID
  }
}

abstract class FluidTankConditions[T <: FluidTankConditions[T]](location: ResourceLocation) extends ICondition {
  final val serializer = new FluidTankConditions.Serializer(this)

  override def getID: ResourceLocation = this.location

  override def test: Boolean
}
